File: backend/scrapers/cmu.py
import requests
from scrapers.uni import Uni
from bs4 import BeautifulSoup

class CMU(Uni):
  def __init__(self):
    super().__init__("cmu")

  def fetch_page(self, url: str) -> str:
    print(f"[{self.name}] Fetching data from {url} using POST...")
    # Payload to fetch CS and ECE courses
    payload = {
        "SEMESTER": "S25", # Spring 2025
        "MINI": "NO",
        "GRAD_UNDER": "All",
        "PRG_LOCATION": "All",
        "DEPT": ["CS", "ECE"],
        "BEG_TIME": "All",
        "KEYWORD": "",
        "TITLE_ONLY": "NO",
        "SUBMIT": "Retrieve Schedule"
    }
    
    try:
        resp = requests.post(url, data=payload, verify=False, timeout=60)
        resp.raise_for_status()
        return resp.text
    except Exception as e:
        print(f"[{self.name}] Error fetching data: {e}")
        return ""

  def links(self):
    return ["https://enr-apps.as.cmu.edu/open/SOC/SOCServlet/search"]

  def parser(self, html):
    soup = BeautifulSoup(html, 'html.parser')
    courses = []
    
    ALLOWED_DEPTS = ['ELECTRICAL & COMPUTER ENGINEERING', 'COMPUTER SCIENCE']
    
    tables = soup.find_all('table', id='search-results-table')
    
    for table in tables:
      # Check department
      prev_h4 = table.find_previous_sibling('h4', class_='department-title')
      if not prev_h4:
        continue
        
      dept_name = prev_h4.get_text(strip=True)
      if dept_name not in ALLOWED_DEPTS:
        continue

      tbody = table.find('tbody')
      if not tbody:
        continue
      
      current_course = None
      
      for tr in tbody.find_all('tr'):
        cols = tr.find_all('td')
        if len(cols) < 10:
          continue
          
        # Extract text from columns
        def get_text(col):
          return col.get_text(strip=True)
        
        course_id_text = get_text(cols[0])
        title_text = get_text(cols[1])
        units_text = get_text(cols[2])
        sec_text = get_text(cols[3])
        days_text = get_text(cols[5])
        begin_text = get_text(cols[6])
        end_text = get_text(cols[7])
        location_text = get_text(cols[8])
        
        # Check if this is a new course
        if course_id_text:
          # Save previous course if it exists
          if current_course:
            courses.append(current_course)
            
          current_course = {
            'id': course_id_text,
            'title': title_text,
            'units': units_text,
            'sections': []
          }
        
        # Parse section info
        meeting = {
          'days': days_text,
          'begin': begin_text,
          'end': end_text,
          'location': location_text
        }
        
        if course_id_text or sec_text:
          # New section
          section = {
            'id': sec_text,
            'meetings': [meeting]
          }
          if current_course:
              current_course['sections'].append(section)
        else:
          # Continuation of previous section
          if current_course and current_course['sections']:
            current_course['sections'][-1]['meetings'].append(meeting)
            
      # Append the last course of this table
      if current_course:
        courses.append(current_course)
      
    return courses

File: src/lib/scrapers/cmu.ts
import * as cheerio from "cheerio";
import { BaseScraper } from "./BaseScraper";
import { Course } from "./types";
import { fetch, Agent } from "undici";

export class CMU extends BaseScraper {
  constructor() {
    super("cmu");
  }

  links(): string[] {
    return ["https://enr-apps.as.cmu.edu/open/SOC/SOCServlet/search"];
  }

  async retrieve(): Promise<Course[]> {
    const url = "https://enr-apps.as.cmu.edu/open/SOC/SOCServlet/search";
    
    const params = new URLSearchParams();
    params.append("SEMESTER", "F25"); // Updated to Fall 2025 to match sample
    params.append("MINI", "NO");
    params.append("GRAD_UNDER", "All");
    params.append("PRG_LOCATION", "All");
    params.append("DEPT", "CS");
    params.append("DEPT", "ECE");
    params.append("BEG_TIME", "All");
    params.append("KEYWORD", "");
    params.append("TITLE_ONLY", "NO");
    params.append("SUBMIT", "Retrieve Schedule");

    const html = await this.fetchPage(url, params);
    if (html) {
      return await this.parser(html);
    }
    return [];
  }

  async fetchPage(url: string, body?: URLSearchParams): Promise<string> {
    console.log(`[${this.name}] Fetching ${url}...`);

    try {
      const response = await fetch(url, {
        method: body ? "POST" : "GET",
        headers: body ? {
          "Content-Type": "application/x-www-form-urlencoded",
        } : {},
        body: body,
        dispatcher: new Agent({
          connect: {
            rejectUnauthorized: false,
          },
        }),
      });

      if (!response.ok) {
        throw new Error(
          `Failed to fetch ${url}: ${response.status} ${response.statusText}`
        );
      }
      return await response.text();
    } catch (error) {
      console.error(`[${this.name}] Error fetching ${url}:`, error);
      return "";
    }
  }

  async fetchDetail(courseCode: string, semester: string): Promise<{ description: string; corequisites: string }> {
    const cleanCode = courseCode.replace(/-/g, "");
    const url = `https://enr-apps.as.cmu.edu/open/SOC/SOCServlet/courseDetails?COURSE=${cleanCode}&SEMESTER=${semester}`;
    
    // Small delay to be polite
    await new Promise(resolve => setTimeout(resolve, 200));

    try {
      const html = await this.fetchPage(url);
      if (!html) return { description: "", corequisites: "" };
      
      const $ = cheerio.load(html);
      const description = $("#course-detail-description p").text().trim();
      
      const prereq = $("dt:contains('Prerequisites')").next("dd").text().trim();
      const coreq = $("dt:contains('Corequisites')").next("dd").text().trim();
      
      let combined = "";
      if (prereq && prereq !== "None") combined += `Prereq: ${prereq}`;
      if (coreq && coreq !== "None") combined += (combined ? "; " : "") + `Coreq: ${coreq}`;
      
      return { description, corequisites: combined };
    } catch (error) {
      console.error(`[${this.name}] Error fetching details for ${courseCode}:`, error);
      return { description: "", corequisites: "" };
    }
  }

  async parser(html: string): Promise<Course[]> {
    const $ = cheerio.load(html);
    const courses: Course[] = [];

    const ALLOWED_DEPTS = [
      "ELECTRICAL & COMPUTER ENGINEERING",
      "COMPUTER SCIENCE",
    ];

    const tables = $("table#search-results-table");
    
    for (const tableElement of tables.toArray()) {
      const table = $(tableElement);
      const prevH4 = table.prevAll("h4.department-title").first();
      if (prevH4.length === 0) continue;

      const deptName = prevH4.text().trim();
      if (!ALLOWED_DEPTS.includes(deptName)) continue;

      const tbody = table.find("tbody");
      if (tbody.length === 0) continue;

      let currentCourse: Course | null = null;
      const rows = tbody.find("tr").toArray();

      for (const trElement of rows) {
        const cols = $(trElement).find("td");
        if (cols.length < 10) continue;

        const getText = (idx: number) => $(cols[idx]).text().trim().replace(/\u00a0/g, " ");
        const rawId = getText(0);

        // CMU course IDs usually look like "15-112" or "15112"
        if (rawId && (/\d{2}-\d{3}/.test(rawId) || /^\d{5}$/.test(rawId))) {
          if (currentCourse) {
            courses.push(currentCourse);
          }

          // Extract URL from onclick attribute if present
          // openModal('#course-detail-modal', '...', '/open/SOC/SOCServlet/courseDetails?COURSE=15050&SEMESTER=F25', ...)
          const onclick = $(cols[0]).find('a').attr('onclick') || '';
          const urlMatch = onclick.match(/openModal\('[^']+',\s*'[^']+',\s*'([^']+)'/);
          let courseUrl = "";
          if (urlMatch) {
            courseUrl = `https://enr-apps.as.cmu.edu${urlMatch[1]}`.replace(/&amp;/g, "&");
          }

          // const { description, corequisites } = await this.fetchDetail(rawId, semester);
          const description = "";
          const corequisites = "";

          // Determine Level: CMU levels are like 15-112. 
          // 100-500 are Undergraduate, 600+ are Graduate.
          let level = "undergraduate";
          const numMatch = rawId.match(/-(\d+)/);
          if (numMatch) {
            const num = parseInt(numMatch[1]);
            if (num >= 600) level = "graduate";
          }

          let department = "Computer Science";
          if (rawId.startsWith("18")) {
            department = "Electrical & Computer Engineering";
          }

          currentCourse = {
            university: this.name,
            courseCode: rawId,
            title: getText(1),
            units: getText(2),
            description: description,
            url: courseUrl,
            department: department,
            corequisites: corequisites,
            level: level,
            semesters: [{ term: "Fall", year: 2025 }],
            details: {
              sections: [],
            },
          };
        }

        if (currentCourse) {
          const secId = getText(3);
          const meeting = {
            days: getText(5),
            begin: getText(6),
            end: getText(7),
            location: getText(8),
          };

          if (rawId || secId) {
            const section = {
              id: secId,
              meetings: [meeting],
            };
            const details = currentCourse.details as { sections: { id: string; meetings: { days: string; begin: string; end: string; location: string }[] }[] };
            details.sections.push(section);
          } else {
            const details = currentCourse.details as { sections: { id: string; meetings: { days: string; begin: string; end: string; location: string }[] }[] };
            const sections = details.sections;
            if (sections.length > 0) {
              sections[sections.length - 1].meetings.push(meeting);
            }
          }
        }
      }
      
      if (currentCourse) {
        courses.push(currentCourse);
      }
    }

    return courses;
  }
}
